### DISCUSSION TITLE
Handling actions in another thread

### DISCUSSION THREAD

**Discussion by edko99** [emilk/egui#484] [Q&A]
My UI has a button that, when clicked, will trigger a heavy computation that will update the app’s state in asynchronous batches. I would like to run this on a separate thread, and have the UI gradually display the updates to the state. Is there a way of doing this?

**Comment by parasyte** [comment]
You are able to update any UI state you want in any way you want. That's the purpose of immediate mode GUIs, after all.

You need to send a `RepaintSignal` to your worker thread so that it can request the GUI to be redrawn when you have new state to show. A `RepaintSignal` can be acquired with the [`repaint_signal()`](https://docs.rs/epi/0.12.0/epi/struct.Frame.html#method.repaint_signal) method on the `Frame` type that your app is given in its `update()` method.

When you are ready to redraw the latest state, just call the [`request_repaint()`](https://docs.rs/epi/0.12.0/epi/trait.RepaintSignal.html#tymethod.request_repaint) method on the `RepaintSignal`.

**Reply by dscso** [reply]
I had some problems with this: here is how I got it working...

```rust
use eframe::egui;
use tokio::sync::mpsc;
use tokio::sync::mpsc::UnboundedReceiver;


#[tokio::main]
async fn main() -> Result<(), eframe::Error> {

    let options = eframe::NativeOptions {
        initial_window_size: Some(egui::vec2(320.0, 240.0)),
        ..Default::default()
    };
    let (tx_gui, mut rx_gui) = mpsc::unbounded_channel();
    eframe::run_native(
        "My egui App",
        options,
        Box::new(|cc| {
            // clone
            let frame = cc.egui_ctx.clone();
            tokio::spawn(async move {
                // sleep async 1 sec
                tokio::time::sleep(tokio::time::Duration::from_secs(3)).await;
                tx_gui.send("Mr Arthur after 3 sec...".to_string()).expect("Failed to send message");
                println!("Repaint send!");
                frame.request_repaint();
            });
            Box::new(MyApp::new(rx_gui))
        }),
    )
}

struct MyApp {
    rx_gui: UnboundedReceiver<String>,
    name: String,
    updates: u32,
}

impl MyApp {
    fn new(rx_gui: UnboundedReceiver<String>) -> Self {
        Self {
            rx_gui,
            name: "Jr. Arthur".to_owned(),
            updates: 0,
        }
    }
}

impl eframe::App for MyApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        self.updates += 1;
        println!("frame update: {}", self.updates);
        if let Ok(msg) = self.rx_gui.try_recv() {
            self.name = msg;
        }
        egui::CentralPanel::default().show(ctx, |ui| {
            ui.heading("My egui Application");
            ui.label(format!("Hello : '{}'", self.name));
            ui.label(format!("Updates: {}", self.updates));
        });
    }
}
```

**Reply by SilenLoc** [reply]
I created a small template (not yet a real template) repo for async handling in [egui](https://github.com/SilenLoc/egui_async_http_call). 
You will see "incoming" inside the AsyncApp that will be mutated every time you press send.
If you want to keep track you can instead add your incoming "thing", here it is a response, to a vector.

Be aware that this example app holds a lot of dependencies with "egui_extras" and "egui_logger"
